cpp/include/cudf/join/filtered_join.hpp

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright (c) 2025, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/table/table_view.hpp>
+#include <cudf/types.hpp>
+#include <cudf/utilities/default_stream.hpp>
+#include <cudf/utilities/export.hpp>
+#include <cudf/utilities/memory_resource.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_uvector.hpp>
+
+#include <utility>
+
+namespace CUDF_EXPORT cudf {
+
+/**
+ * @addtogroup column_join
+ * @{
+ * @file
+ */
+
+namespace detail {
+/**
+ * @brief Forward declaration for our filtered hash join
+ */
+class filtered_join;
+}  // namespace detail
+
+/**
+ * @brief Filtered hash join that builds hash table in creation and probes results in subsequent
+ * `*_join` member functions
+ *
+ * This class enables the filtered hash join scheme that builds hash table once, and probes as many
+ * times as needed (possibly in parallel).
+ *
+ * @note All NaNs are considered as equal
+ */
+class filtered_join {
+ public:
+  filtered_join() = delete;
+  ~filtered_join();
+  filtered_join(filtered_join const&)            = delete;
+  filtered_join(filtered_join&&)                 = delete;
+  filtered_join& operator=(filtered_join const&) = delete;
+  filtered_join& operator=(filtered_join&&)      = delete;
+
+  /**
+   * @brief Constructs a filtered hash join object for subsequent probe calls
+   *
+   * @param build The build table
+   * @param compare_nulls Controls whether null join-key values should match or not
+   * @param reuse_left_table Boolean indicating if the build table should be reused. If true, then
+   * the build table is considered as the left table, and is reused with multiple right (probe)
+   * tables. If false, then the build table is the right/filter table, it will be applied to
+   * multiple left (probe) tables.
+   * @param stream CUDA stream used for device memory operations and kernel launches
+   */
+  filtered_join(cudf::table_view const& build,
+                cudf::null_equality compare_nulls = null_equality::EQUAL,
+                bool reuse_left_table             = false,
```

Comment:
we probably want to introduce a strong type if it's a public interface, e.g. `reuse_left_table::YES/NO`

Reply:
This variable name is also confusing. This means "we want to reuse only the left table from the two tables", while the docs say this means "we will use the table as the left table and reuse it". Since the table is always reused, probably we should call it `build_as_left_table`.